exchanges/dydx/dydx_websocket.go

```diff
@@ -0,0 +1,328 @@
+	defer dy.Websocket.Wg.Done()
+	for {
+		resp := dy.Websocket.Conn.ReadMessage()
+		if resp.Raw == nil {
+			return
+		}
+		err := dy.wsHandleData(resp.Raw)
+		if err != nil {
+			dy.Websocket.DataHandler <- err
+		}
+	}
+}
+
+func (dy *DYDX) wsHandleData(respRaw []byte) error {
+	var resp WsResponse
+	err := json.Unmarshal(respRaw, &resp)
+	if err != nil {
+		return err
+	}
+	switch resp.Channel {
+	case accountsChannel:
+		switch resp.Type {
+		case "subscribe":
+			var resp AccountSubscriptionResponse
+			err = json.Unmarshal(respRaw, &resp)
+			if err != nil {
+				return err
+			}
+			err = dy.processOrders(resp.Contents.Orders)
+			if err != nil {
+				return err
+			}
+			dy.Websocket.DataHandler <- resp.Transfers
+			dy.Websocket.DataHandler <- resp.FundingPayments
+		case "channel_data":
+			var resp AccountChannelData
+			err := json.Unmarshal(respRaw, &resp)
+			if err != nil {
+				return err
+			}
+			err = dy.processOrders(resp.Contents.Orders)
+			if err != nil {
+				return err
+			}
+			dy.Websocket.DataHandler <- resp.Contents.Accounts
+			dy.Websocket.DataHandler <- resp.Contents.Positions
+			dy.Websocket.DataHandler <- resp.Contents.Fills
+		}
+		return nil
+	case orderbookChannel:
+		var market MarketOrderbook
+		err := json.Unmarshal(respRaw, &market)
+		if err != nil {
+			return err
+		}
+		pair, err := currency.NewPairFromString(resp.ID)
+		if err != nil {
+			return err
+		}
+		newOrderbook := orderbook.Base{
+			Asset:       asset.Spot,
+			Asks:        market.Asks.generateOrderbookItem(),
+			Bids:        market.Bids.generateOrderbookItem(),
+			Exchange:    dy.Name,
+			Pair:        pair,
+			LastUpdated: time.Now(),
+		}
+		err = dy.Websocket.Orderbook.LoadSnapshot(&newOrderbook)
+		if err != nil {
+			return err
+		}
+		return nil
+	case tradesChannel:
+		var myTrades MarketTrades
+		err := json.Unmarshal(resp.Contents, &myTrades)
+		if err != nil {
+			return err
+		}
+		pair, err := currency.NewPairFromString(resp.ID)
+		if err != nil {
+			return err
+		}
+		trades := make([]trade.Data, len(myTrades.Trades))
+		for i := range myTrades.Trades {
+			side, err := order.StringToOrderSide(myTrades.Trades[i].Side)
+			if err != nil {
+				return err
+			}
+			trades[i] = trade.Data{
+				Amount:       myTrades.Trades[i].Size,
+				AssetType:    asset.Spot,
+				CurrencyPair: pair,
+				Exchange:     dy.Name,
+				Side:         side,
+				Timestamp:    time.Now(),
+				Price:        myTrades.Trades[i].Price,
+			}
+		}
+		return trade.AddTradesToBuffer(dy.Name, trades...)
+	case marketsChannel:
+		var market InstrumentDatas
+		err := json.Unmarshal(resp.Contents, &market)
+		if err != nil {
+			return err
+		}
+		tickers := make([]ticker.Price, len(market.Markets))
+		count := 0
+		for x := range market.Markets {
+			pair, err := currency.NewPairFromString(x)
+			if err != nil {
+				return err
+			}
+			tickers[count] = ticker.Price{
+				ExchangeName: dy.Name,
+				Ask:          market.Markets[x].IndexPrice,
```

Comment:
This seems like you are using the key to access a map, it's more efficient to declare a variable above.